### ISSUE TITLE
Question: VR vs Raft durability

### ISSUE THREAD

**Issue by freeekanayaka** [tigerbeetle/tigerbeetle#1783]
Hello,

first of all, thanks for this great project! Using an issue to ask this, please let me know if there's a better place/way.

With Raft, when a client sends a command to the leader and then gets back a successful response, it is guaranteed that the command was committed and that it will not be lost even if all nodes that received the command crash right after the response. That's because Raft requires the command to be persisted on disk.

With Viewstamped Replication, the guarantee seems slightly weaker, as far as I understand it. If a client sends an operation to the primary, and then gets back a successful reply, it is still possible for the operation to be lost, for example in case all replicas that got notified about the operation crash before persisting that operation to disk.

So in order for TigerBeetle to operate safely, it is required that replicas that have received a certain committed operation never all crash before persisting it.

Since there's no way to guarantee that, it should be theoretically possible for TigerBeetle to reply to a client that a request was successful, but then lose it because some catastrophic failure.

Is my understanding correct?

This is mostly an academic question, since I understand that it's reasonable to say that those scenarios don't happen in practice. After all, the same would happen in a Raft-based implementation if all the servers that received a command crash AND their disks fail in a way that data can't be recovered (although this scenario is strictly speaking less likely then just the crash-before-persisting one).

**Comment by sentientwaffle** [comment]
Great question! There's two different answers:

1. How it works in Viewstamped Replication (VR) (as described in [Viewstamped Replication Revisited](https://pmg.csail.mit.edu/papers/vr-revisited.pdf)).
2. And how it works in TigerBeetle.

---

VR does not require the disk at all. The protocol is described as if all state exists in memory.

However, VR _does_ guarantee that operations are never lost due to a crash. It accomplishes this with "recovery protocol" (§4.3), which uses consensus to bring crashed-then-restarted replicas safely back into the cluster. In other words, if a replica crashes and restarts, before it can rejoin consensus (i.e. before the restarted replica can ack or commit any prepares) it must complete Recovery Protocol, with the help of `f+1` healthy replicas. (See §8.2).

This means that a VR(R) cluster can recover from at most `f` (out of `2f+1`) unhealthy replicas! If `f+1` replicas fail, then the cluster would be permanently unavailable. This is a tradeoff it makes: without disk, it can't recover, but it also isn't subject to disk faults.

---

Instead of recovery protocol, TigerBeetle uses the disk:

- Before participating in a particular view, a replica persists its intent to participate in that view to the ["superblock"](https://docs.tigerbeetle.com/internals/data_file). This ensures that if the replica restarts, it can never backtrack to an old view.
- And to ensure that committed messages are not lost, we rely on [Protocol-Aware Recovery](https://www.usenix.org/system/files/conference/fast18/fast18-alagappan.pdf) of the write-ahead log.

(See #426 for the change that removed Recovery Protocol from TigerBeetle in favor of the superblock.)

To summarize: Even if all replicas crash, TigerBeetle's VR implementation recovers with no data loss.

(And if there is catastrophic simultaneous corruption of the same operation on all 6 replica's disks, then TigerBeetle will safely shut down).

**Comment by freeekanayaka** [comment]
Thanks for the quick reply!

I changed the title of the issue to emphasize this is to understand durability rather than availability.

---

Regarding VR: yes, it's clear to me that a single crashed replica can recover data loss using the recovery protocol. However let's say that we have a cluster with 3 replicas and the following happens:

1. Client C sends an operation request to replica R1, which is the primary
2. Replica R1 replicates the request to R2, which acks it
3. Replica R1 replies to C saying that the operation is committed
4. Replica R3 is partitioned from R1 and R2, and does not receive the operation
5. Both R1 and R2 crash
6. The cluster is now unavailable
7. R1 and R2 restart, and connectivity is also restored with R3, the cluster is again fully available

If I understand VR correctly, in this specific case the committed operation will be lost permanently **even if it was already acknowledged to the client**.

Of course there are more than `f` replicas failing in this scenario (where `f = 1`). My point is that in this case VR does not only fail to guarantee availability, but also **durability**.

Contrast that with Raft, where the same scenario will likewise fail to guarantee availability but it will **not** fail to guarantee durability, since in Raft a majority of servers must persist the operation before it's considered committed.

---

Regarding TigerBeetle, I am not yet familiar with Protocol-Aware Recovery, however I was under the impression that TigerBeetle does not wait for operations to be persisted before considering them committed and replying to the client, and so it can't guarantee durability in the scenario above, regardless of whether recovery is implemented with the regular VR recovery protocol or with Protocol-Aware Recovery.

I'm not entirely sure about this though, since I don't know the details of the TigerBeetle implementation.

**Comment by matklad** [comment]
> If I understand VR correctly, in this specific case the committed operation will be lost permanently even if it was already acknowledged to the client.

To spell this out, in this scenario the Viewstamp Replication Revisited remains permanintelly unavailable. 

After the crash, for R1 and R2 to recover, they need 2 RecoveryResponse messages. But only normal replicas can send a RecoveryResponse, and only R3 is normal. So, the cluster remains forever unavailable. 

Which means that you can't read _anything_ from the cluster, because a read operation also need to go through quorum, and we don't have one. 

This is indeed a weakness of VRR --- f+1 simultaneous failures make the cluster _permanently_ unavailable. 

Raft (and the modified VRR used by TigerBeetle) allow for the number of faults to _temporarily_ dip beyond f, and guarantee that availability is restored when the number of faults goes back bellow `f`.

Another way to phrase this is that, in VRR, if the entire cluster crashes simultaneously, then all data is lost (pure VRR doesn't store anything to disk at all).

>however I was under the impression that TigerBeetle does not wait for operations to be persisted 

This is not correct. We do wait for data to be durably appended to the log before acknowledging it. Here's where we durably save a message:

https://github.com/tigerbeetle/tigerbeetle/blob/cd2c8b2ebd404d41bcd43bd40daa751830966a2d/src/vsr/replica.zig#L1626

and here's where we send an acknowledgement message (prepare_ok), _after_ the prepare was persisted:

https://github.com/tigerbeetle/tigerbeetle/blob/cd2c8b2ebd404d41bcd43bd40daa751830966a2d/src/vsr/replica.zig#L9205

---

> With Raft, when a client sends a command to the leader and then gets back a successful response, it is guaranteed that the command was committed and that it will not be lost even if all nodes that received the command crash right after the response. That's because Raft requires the command to be persisted on disk.

An important asterisk here is that this is true under the assumption that the disk is infallible.  That is, that that's impossible for the disk to acknowledge the write and then fail to fulfill the read after restart. 

TigerBeetle goes slightly beyond that, and guarantees durability even if some storage operations fail.

**Comment by freeekanayaka** [comment]
> > If I understand VR correctly, in this specific case the committed operation will be lost permanently even if it was already acknowledged to the client.
> 
> To spell this out, in this scenario the Viewstamp Replication Revisited remains permanintelly unavailable.
> 
> After the crash, for R1 and R2 to recover, they need 2 RecoveryResponse messages. But only normal replicas can send a RecoveryResponse, and only R3 is normal. So, the cluster remains forever unavailable.

Thanks for pointing out this. It's of course much better to remain permanently unavailable than to silently lose commits. I'm very familiar with Raft, but I still have to read the VRR paper through and digest it, so this is a valuable insight, and mitigates my concern around it a lot.

I'm still not very clear about what happens if one tries to optimize VRR by storing operations to disk asynchronously (out-of-band from the core protocol, without waiting these writes to complete/succeed) so recovery can be faster after a crash. Does your argument still hold? Is there always a way to guarantee that the cluster will just become permanently unavailable?

Anyway, I'll make sure to get deeper into VRR first.

> Which means that you can't read _anything_ from the cluster, because a read operation also need to go through quorum, and we don't have one.
> 
> This is indeed a weakness of VRR --- f+1 simultaneous failures make the cluster _permanently_ unavailable.

Gotcha, that's a great way to put it. I think it's kind of a reasonable tradeoff in exchange for diskless operation.
 
> Raft (and the modified VRR used by TigerBeetle) allow for the number of faults to _temporarily_ dip beyond f, and guarantee that availability is restored when the number of faults goes back bellow `f`.
> 

Exactly.

> Another way to phrase this is that, in VRR, if the entire cluster crashes simultaneously, then all data is lost (pure VRR doesn't store anything to disk at all).
> 

Right. I'd like to understand what happens in a hybrid implementation were data is stored on disk as optimization.

> > however I was under the impression that TigerBeetle does not wait for operations to be persisted
> 
> This is not correct. We do wait for data to be durably appended to the log before acknowledging it. Here's where we durably save a message:
> 
> https://github.com/tigerbeetle/tigerbeetle/blob/cd2c8b2ebd404d41bcd43bd40daa751830966a2d/src/vsr/replica.zig#L1626
> 
> and here's where we send an acknowledgement message (prepare_ok), _after_ the prepare was persisted:
> 
> https://github.com/tigerbeetle/tigerbeetle/blob/cd2c8b2ebd404d41bcd43bd40daa751830966a2d/src/vsr/replica.zig#L9205
> 

Oh I see. From the videos I had watched, I thought TigerBeetle was taking advantage of the diskless nature of VRR and **not** waiting for data to be persisted (which is not required for correctness of VRR, while it is for correctness of Raft).

> > With Raft, when a client sends a command to the leader and then gets back a successful response, it is guaranteed that the command was committed and that it will not be lost even if all nodes that received the command crash right after the response. That's because Raft requires the command to be persisted on disk.
> 
> An important asterisk here is that this is true under the assumption that the disk is infallible. That is, that that's impossible for the disk to acknowledge the write and then fail to fulfill the read after restart.

Yes, sure. That's what I meant at the end of my first comment: if all Raft servers that participated in a commit crash **and** all their disks lose that commit, then the situation is of course the same as VRR.
 
> TigerBeetle goes slightly beyond that, and guarantees durability even if some storage operations fail.

Yeah, I gathered that, that's indeed very interesting and something I'll dig deeper into.

**Comment by jorangreef** [comment]
@freeekanayaka since last you opened this issue, we did a talk in June in NYC at Systems Distributed '24 on Durability in TigerBeetle and how we use VSR to exchange (replicated) durability for availability, plus maximize the efficiency of this exchange. Also, some interesting durability failure modes we've discovered that TB can handle, but which an off-the-shelf implementation of Paxos or Raft without an explicit storage fault model might probably not.

In case you find it interesting:

https://www.youtube.com/watch?v=tRgvaqpQPwE

**Comment by freeekanayaka** [comment]
Thank you very much for the pointer, keep up the good work!
